InterfaceFortran/TasmanianSG.f90

```diff
@@ -154,501 +169,732 @@ Module TasmanianSG
+
+  opt_flags = 0
+  if (present(output)) then
+    opt_flags(1) = 1
+    theout = output-1
+  endif
+  if (present(levelLimits)) then
+    opt_flags(2) = 1
+    ll => levelLimits
+  endif
+
+  call tsgshr(gridID, tolerance, rtype, opt_flags, theout, ll)
+end subroutine tsgSetLocalSurplusRefinement
+!=======================================================================
+subroutine tsgClearRefinement(gridID)
+  integer :: gridID
+  call tsgcre(gridID)
+end subroutine tsgClearRefinement
+!=======================================================================
+subroutine tsgMergeRefinement(gridID)
+  integer :: gridID
+  call tsgmre(gridID)
+end subroutine tsgMergeRefinement
+!=======================================================================
+subroutine tsgSetConformalTransformASIN(gridID, truncate)
+  integer :: gridID, truncate(:)
+  call tsgsca(gridID, truncate)
+end subroutine tsgSetConformalTransformASIN
+!=======================================================================
+function tsgIsSetConformalTransformASIN(gridID) result(isset)
+  integer :: gridID, res
+  logical :: isset
+  call tsgica(gridID, res)
+  if (res .EQ. 0)then
     isset = .FALSE.
-  ELSE
+  else
     isset = .TRUE.
-  ENDIF
-END FUNCTION tsgIsSetConformalTransformASIN
+  endif
+end function tsgIsSetConformalTransformASIN
 !=======================================================================
-SUBROUTINE tsgClearConformalTransform(gridID)
-  INTEGER :: gridID
-  CALL tsgcct(gridID)
-END SUBROUTINE tsgClearConformalTransform
+subroutine tsgClearConformalTransform(gridID)
+  integer :: gridID
+  call tsgcct(gridID)
+end subroutine tsgClearConformalTransform
 !=======================================================================
-FUNCTION tsgGetConformalTransformASIN(gridID) result(truncate)
-  INTEGER :: gridID, N
-  INTEGER, allocatable :: truncate(:)
+function tsgGetConformalTransformASIN(gridID) result(truncate)
+  integer :: gridID, N
+  integer, allocatable :: truncate(:)
   N = tsgGetNumDimensions(gridID)
-  ALLOCATE(truncate(N))
-  CALL tsgcre(gridID, truncate)
-END FUNCTION tsgGetConformalTransformASIN
-!=======================================================================
-SUBROUTINE tsgPrintStats(gridID)
-  INTEGER :: gridID
-  CALL tsgpri(gridID)
-END SUBROUTINE tsgPrintStats
-!=======================================================================
-! DO NOT CALL THOSE FUNCTIONS DIRECTLY !
-!=======================================================================
-SUBROUTINE tsgReceiveString(l, S)
-  INTEGER :: l
-  CHARACTER, target :: S(l)
+  allocate(truncate(N))
+  call tsgcre(gridID, truncate)
+end function tsgGetConformalTransformASIN
+!=======================================================================
+subroutine tsgPrintStats(gridID)
+  integer :: gridID
+  call tsgpri(gridID)
+end subroutine tsgPrintStats
+!=======================================================================
+function tsgTestInternals(verbose) result(res)
```

Comment:
* If I read this correctly, the only way this test can pass is if it gets called before any `tsgNewGridID()`, is that correct?
* You can make the first command of the test to be `tsgClearAll()` and add comment that this will delete all grids.

Reply:
* This is true
* Instead I can call tsggag(num_ag) first and use it as a base value

Reply:
* Fix it either way, use a base value or call `tsgClearAll()` with an added comment

Reply:
However, we test tsgClearAll() at the end anyway

Reply:
You decide, `tsgClearAll()` should be called at least once, and `tsgTestInternals()` should either not-crash or add comment about why it is crashing.

Reply:
Also, current implementation of tsgEvaluateHierarchicalFunctions() allows one to call it on Fourier grid. 
The result is real array with twice as many elements.
Do you want me to change this?

Reply:
Yes, if we have complex evaluate, there is no need to keep the other one.